### PR TITLE
api: support errors extended information and error type in MessagePack

### LINKED ISSUE #209: Support the ERROR type
See the feature description:

https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/#the-error-type

### PR DESCRIPTION
**Pull request by DifferentialOrange**
Merge after #226

### code health: rename error iproto code

Since 2.4.1 IPROTO_ERROR (0x31) is renamed to IPROTO_ERROR_24 and IPROTO_ERROR name is used for 0x52 constant describing extended error info [1].

1. https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/#the-error-type

Part of #209

### api: support errors extended information

Since Tarantool 2.4.1, iproto error responses contain extended info with backtrace [1]. After this patch, Error would contain ExtendedInfo field (BoxError object), if it was provided. Error() handle now will print extended info, if possible.

1. https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/#responses-for-errors

Part of #209

### api: support error type in MessagePack

Tarantool supports error extension type since version 2.4.1 [1], encoding was introduced in Tarantool 2.10.0 [2]. This patch introduces the support of Tarantool error extension type in msgpack decoders and encoders.

Tarantool error extension type objects are decoded to `*tarantool.BoxError` type. `*tarantool.BoxError` may be encoded top Tarantool error extension type objects.

Error extension type internals are the same as errors extended information: the only difference is that extra information is encoded as a separate error dictionary field and error extension type objects are encoded as MessagePack extension type objects.

The only way to receive an error extension type object from Tarantool is to receive an explicitly built `box.error` object: either from
`return box.error.new(...)` or a tuple with it. All errors raised within Tarantool (including those raised with `box.error(...)`) are encoded based on the same rules as simple errors due to backward compatibility.

It is possible to create error extension type objects with Go code, but it not likely to be really useful since most of their fields is
computed on error initialization on the server side (even for custom error types).

This patch also adds ErrorExtensionFeature flag to client protocol features list. Without this flag, all `box.error` object sent over
iproto are encoded to string. We behave like Tarantool `net.box` here: if we support the feature, we provide the feature flag.

Since it may become too complicated to enable/disable feature flag through import, error extension type is available as a part of the base package, in contrary to Decimal, UUID, Datetime and Interval types which are enabled by importing underscore subpackage.

1. tarantool/tarantool#4398
2. tarantool/tarantool#6433

Closes #209

I didn't forget about:

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

### CODE REVIEW

File: box_error.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+func (err BoxError) Depth() (depth int) {
+	depth = 1
+	for err.Prev != nil {
+		err = *err.Prev
+		depth++
+	}
+	return depth
+}
```
Comment (oleg-jukovec) [inline box_error.go:32]: 1. It would be nice to use this type as common errors. To do this, you need to implement the `error` interface.

https://go.dev/tour/methods/19
```Go
type error interface {
    Error() string
}
```

2. It would be great if the type was field-compatible (in filed-naming) with existing ones:
https://github.com/tarantool/go-tarantool/blob/master/errors.go#L5-L21
`Errcode -> Code`, `Message` -> `Msg`

Comment (DifferentialOrange) [reply box_error.go]: Updated both

File: tarantool_test.go [hunk 2]
```diff
@@ -3150,6 +3150,91 @@ func TestConnectionFeatureOptsImmutable(t *testing.T) {
...
+	_, err := conn.Eval("error(access_denied_error)", []interface{}{})
+	require.NotNilf(t, err, "expected error on forbidden action")
+
+	ttErr, ok := err.(Error)
+	require.Equalf(t, ok, true, "error is built from a Tarantool error")
+
+	expected := BoxError{
+		Type:  "AccessDeniedError",
+		File:  "/__w/sdk/sdk/tarantool-2.10/tarantool/src/box/func.c",
+		Line:  uint32(535),
```
Comment (oleg-jukovec) [inline tarantool_test.go:3232]: The line check will fail on new Tarantool versions.

File: box_error.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+				return nil, err
+			}
+		}
+	}
+
+	if len(errorStack) > 0 {
+		return &errorStack[0], nil
+	}
+
+	return nil, nil
```
Comment (oleg-jukovec) [inline box_error.go:150]: Is there no error?

File: box_error_msgpack_test.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
+package tarantool_test
```
Comment (oleg-jukovec) [inline box_error_msgpack_test.go:1]: In general, it could be better tested, although it is meditative.

https://coveralls.io/builds/54505878/source?filename=box_error.go

File: box_error.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+func (boxError *BoxError) UnmarshalMsgpack(b []byte) error {
+	var buf smallBuf = smallBuf{b: b}
+
+	dec := newDecoder(&buf)
+	val, err := decodeBoxError(dec)
+	if err != nil {
+		return err
+	}
+
+	*boxError = *val
```
Comment (oleg-jukovec) [inline box_error.go:271]:
```suggestion
        if val, err := decodeBoxError(dec) err != nil {
            return err
        } else {
            *boxError = *val
        }
```

File: box_error.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+		return b, err
+	}
+
+	return buf.b, nil
+}
+
+// UnmarshalMsgpack deserializes a BoxError value from a MessagePack
+// representation.
+func (boxError *BoxError) UnmarshalMsgpack(b []byte) error {
+	var buf smallBuf = smallBuf{b: b}
```
Comment (oleg-jukovec) [inline box_error.go:263]: It can be just a byte buffer. There is no need in smallBuf (which may be confusing).
```suggestion
	dec := newDecoder(bytes.NewBuffer(b))
```

File: box_error.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+			boxError = boxError.Prev
+		}
+	}
+
+	return nil
+}
+
+// MarshalMsgpack serializes the BoxError into a MessagePack representation.
+func (boxError *BoxError) MarshalMsgpack() (b []byte, err error) {
+	var buf smallWBuf
```
Comment (oleg-jukovec) [inline box_error.go:250]: The same about smallWBuf.

File: box_error.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	return depth
+}
+
+func decodeBoxError(d *decoder) (*BoxError, error) {
+	var l, larr, l1, l2 int
+	var errorStack []BoxError
+	var err error
+	var mapk, mapv interface{}
```
Comment (oleg-jukovec) [inline box_error.go:60]: You can declare variables closer to a usege places.

File: box_error.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	// Fields are additional fields depending on error type. For example, if
+	// type is "AccessDeniedError", then it will include "object_type",
+	// "object_name", "access_type".
+	Fields map[interface{}]interface{}
+	// Prev is the previous error in stack.
+	Prev *BoxError
+}
+
+// Error converts a BoxError to a string.
+func (boxError *BoxError) Error() string {
```
Comment (oleg-jukovec) [inline box_error.go:36]: https://github.com/golang/go/wiki/CodeReviewComments#receiver-names

```suggestion
func (e *BoxError) Error() string {
```

File: box_error.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	if len(errorStack) > 0 {
+		return &errorStack[0], nil
+	}
+
+	return nil, nil
+}
+
+func encodeBoxError(enc *encoder, boxError *BoxError) (err error) {
```
Comment (oleg-jukovec) [inline box_error.go:153]: This is not a very good practice, although unfortunately it occurs in our code:

```suggestion
func encodeBoxError(enc *encoder, boxError *BoxError) error {
```

File: box_error.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+		if stackDepth > 1 {
+			boxError = boxError.Prev
+		}
+	}
+
+	return nil
+}
+
+// MarshalMsgpack serializes the BoxError into a MessagePack representation.
+func (boxError *BoxError) MarshalMsgpack() (b []byte, err error) {
```
Comment (oleg-jukovec) [inline box_error.go:249]:
```suggestion
func (boxError *BoxError) MarshalMsgpack() ([]byte, error) {
```

File: box_error_msgpack_test.go [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+	"github.com/stretchr/testify/require"
+	. "github.com/tarantool/go-tarantool"
+	"github.com/tarantool/go-tarantool/test_helpers"
+)
+
+var space = "test_error_type"
+var index = "primary"
+
+type TupleBoxError struct {
+	pk  string // BoxError cannot be used as primary key
```
Comment (oleg-jukovec) [inline box_error_msgpack_test.go:16]:
```suggestion
	pk  string // BoxError cannot be used as a primary key.
```

**Review by oleg-jukovec — COMMENTED** [review]
Thanks for the patch. In general, everything is fine, except for a couple of few codestyle moments.

File: tarantool_test.go (hunk 2 above)
Comment (DifferentialOrange) [reply tarantool_test.go:3232]: Yeah, so test error comparison ignores them

File: box_error.go [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+	// File is a source code file where the error was caught.
+	File string
+	// Line is a number of line in the source code file where the error was caught.
+	Line uint32
+	// Msg is the text of reason.
+	Msg string
+	// Errno is the ordinal number of the error.
+	Errno uint32
+	// Code is the number of the error as defined in `errcode.h`.
+	Code uint32
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply box_error.go:26]: Well, unfortunately, it is a tricky one.

Tarantool uses uint32 to store these fields. msgpack v5 (not v2) encodes uint64 as honest uint64 values, and when Tarantool tries to encode a msgpack payload with uint64-filled error, it fails: https://github.com/tarantool/tarantool/issues/7968

File: const.go [hunk 14]
```diff
@@ -56,6 +57,15 @@ const (
...
 	KeySQLInfoAutoincrementIds = 0x01
 
+	KeyErrorStack   = 0x00
+	KeyErrorType    = 0x00
+	KeyErrorFile    = 0x01
+	KeyErrorLine    = 0x02
+	KeyErrorMessage = 0x03
+	KeyErrorErrno   = 0x04
+	KeyErrorErrcode = 0x05
+	KeyErrorFields  = 0x06
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply const.go:67]: Well, no, but similar codes for other request internals are public constants too:
https://github.com/tarantool/go-tarantool/blob/7e86795514b0e0637372bee473deb8eaefb51401/const.go#L22-L45

File: box_error.go (hunk 3 above)
Comment (DifferentialOrange) [reply box_error.go:150]: To be honest, I don't know how to treat this one. MP_ERROR type documentation does not declare any restrictions on stack size. On the other hand, I can't imagine how error without any errors in stack may exist.

File: box_error_msgpack_test.go (hunk 4 above)
Comment (DifferentialOrange) [reply box_error_msgpack_test.go:1]: For now there are no similar tests in common-purpose decoding:

https://coveralls.io/builds/54545306/source?filename=response.go

But I can try to make it if you wish.

File: box_error.go (hunk 6 above)
Comment (DifferentialOrange) [reply box_error.go:263]: I had inherited this from common-purpose Marshal/Unmarshal, but it seems that plain buffer should be enough too. (But a little consistency sounds nice too.)

File: box_error.go (hunk 5 above)
Comment (DifferentialOrange) [reply box_error.go]: Linter doesn't like else after return (at least with my local config), implemented the first half.

File: box_error.go (hunk 6 above)
Comment (DifferentialOrange) [reply box_error.go]: Changed

File: box_error.go (hunk 7 above)
Comment (DifferentialOrange) [reply box_error.go]: Changed

File: box_error.go (hunk 8 above)
Comment (DifferentialOrange) [reply box_error.go]: Moved

File: box_error.go (hunk 11 above)
Comment (DifferentialOrange) [reply box_error.go]: Changed

File: box_error_msgpack_test.go (hunk 12 above)
Comment (DifferentialOrange) [reply box_error_msgpack_test.go]: Fixed

File: box_error.go (hunk 9 above)
Comment (DifferentialOrange) [reply box_error.go]: Changed

File: test_helpers/utils.go [hunk 15]
```diff
@@ -112,3 +113,71 @@ func SkipIfIdSupported(t *testing.T) {
...
+// ...
+//
+// Tarantool error file and line could differ even between
+// different patches.
+//
+// So we check equivalence of all attributes except for Line and File.
+// For Line and File, we check that they are filled with some non-default value
+// (since lines are counted starting with 1 and empty file path is not expected
+// too).
+func EqualBoxErrors(t *testing.T, expected tarantool.BoxError, actual tarantool.BoxError) {
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply test_helpers/utils.go]: Changed

File: box_error.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply box_error.go]: Well, it seems that it is ok to use uint64 with v5 (while using int64 with v5 fails and with v2 is ok). But I'm still not sure which one we should pick. Implemented uint64 version for now.

File: box_error.go [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+	// Fields are additional fields depending on error type. For example, if
+	// type is "AccessDeniedError", then it will include "object_type",
+	// "object_name", "access_type".
+	Fields map[interface{}]interface{}
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply box_error.go]: Changed

File: box_error.go [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+	// Prev is the previous error in stack.
+	Prev *BoxError
+}
+
+// Error converts a BoxError to a string.
+func (boxError *BoxError) Error() string {
+	return fmt.Sprintf("%s (%s, code 0x%x), see %s:%d",
+		boxError.Msg, boxError.Type, boxError.Code,
+		boxError.File, boxError.Line)
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply box_error.go]: Added recursive print for `Prev` and unit test.

To be honest, I think `Fields` doesn't worth adding to. For example, the info stored in AccessDeniedError fields is a duplication of info already mentioned in the message. Like, it may be convenient if you want to inspect in code which exact user was rejected without parsing the message string, but it is rather useless in error printout. Same goes for constraint violation errors, see [field definitions](https://github.com/tarantool/tarantool/blob/c60a820506322419c9537c0cd641b13cc8999e0c/src/box/tuple_constraint_func.c#L104-L113) and [message example](https://github.com/tarantool/tarantool/blob/ed9b982d319ae4020e0536333c748526287a27f5/test/engine-luatest/gh_6436_field_constraint_test.lua#L96).

File: box_error_msgpack_test.go (hunk 4 above)
Comment (DifferentialOrange) [reply box_error_msgpack_test.go]: Added tests for possible msgpack decode corner cases. Has zero idea how to cover possible encode errors. (I think the only way to fail this is to mess with buffer size, which is purely internal.)

**Comment by DifferentialOrange** [comment]
All comments are either resolved or responded, it's time for the next round.

File: const.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply const.go:67]: It's more of a mistake: https://github.com/tarantool/go-tarantool/issues/158

**Review by oleg-jukovec — COMMENTED** [review]
Let's do rebase.

File: const.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply const.go]: Changed for these keys, do not changed for `IPROTO_ERROR`/`IPROTO_ERROR_24` for the consistency.

File: box_error.go [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+
+	if val != nil {
+		if e == nil {
+			e = &BoxError{}
+		}
+		*e = *val
+	}
+
+	return nil
+}
```
Comment (oleg-jukovec) [inline box_error.go:277]: It will be strange to have an error, but with default fields.
```suggestion
func (e *BoxError) UnmarshalMsgpack(b []byte) error {
	buf := bytes.NewBuffer(b)
	dec := newDecoder(buf)

	if val, err := decodeBoxError(dec); err != nil {
		return err
	} else if val == nil {
		return fmt.Errorf("MP_ERROR expected, but not found")
	} else {
		*e = *val
		return nil
	}
}
```

File: box_error.go [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+	dec := newDecoder(buf)
+
+	if val, err = decodeBoxError(dec); err != nil {
+		return err
+	}
+
+	if val != nil {
+		if e == nil {
+			e = &BoxError{}
+		}
```
Comment (oleg-jukovec) [inline box_error.go:272]: It doesn't seem to make sense. Will we set a value to a stack-local varibale at the next line?

File: box_error.go [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+	return nil
+}
+
+// MarshalMsgpack serializes the BoxError into a MessagePack representation.
+func (e *BoxError) MarshalMsgpack() ([]byte, error) {
+	var buf bytes.Buffer
+
+	enc := newEncoder(&buf)
+	if err := encodeBoxError(enc, e); err != nil {
+		return []byte{}, err
```
Comment (oleg-jukovec) [inline box_error.go:285]: nil is a valid value for an empty slice:
```suggestion
		return nil, err
```

File: box_error_test.go [hunk 21]
```diff
@@ -0,0 +1,466 @@
...
+}
+
+func TestErrorTypeMPEncodeDecode(t *testing.T) {
+	for name, testcase := range tupleCases {
+		t.Run(name, func(t *testing.T) {
+			var buf []byte
+			var err error
+
+			buf, err = marshal(&testcase.tuple)
+			require.Nil(t, err)
```
Comment (oleg-jukovec) [inline box_error_test.go:271]:
```suggestion
			buf, err := marshal(&testcase.tuple)
			require.Nil(t, err)
```

File: const.go [hunk 22]
```diff
@@ -56,6 +57,15 @@ const (
...
 	KeySQLInfoAutoincrementIds = 0x01
 
+	keyErrorStack   = 0x00
+	keyErrorType    = 0x00
+	keyErrorFile    = 0x01
+	keyErrorLine    = 0x02
+	keyErrorMessage = 0x03
+	keyErrorErrno   = 0x04
+	keyErrorErrcode = 0x05
+	keyErrorFields  = 0x06
```
Comment (oleg-jukovec) [inline const.go:67]: What about move this constants to box_error.go? These constants don't seem to be used anywhere else.

File: errors.go [hunk 23]
```diff
@@ -4,12 +4,17 @@ import "fmt"
 
 // Error is wrapper around error returned by Tarantool.
 type Error struct {
-	Code uint32
-	Msg  string
+	Code         uint32
+	Msg          string
+	ExtendedInfo *BoxError
```
Comment (oleg-jukovec) [inline errors.go:9]: err.Code == err.ExtendedInfo.Code?
err.Msg == err.ExtendedInfo.Msg?

Does it make sense?

```Go
type Error struct {
    Code uint32
    Msg   string
}

type BoxError struct {
    Error
    // other fields
}
```

?

File: tarantool_test.go [hunk 24]
```diff
@@ -3148,6 +3156,91 @@ func TestConnectionFeatureOptsImmutable(t *testing.T) {
...
+	_, err := conn.Eval("error(access_denied_error)", []interface{}{})
+	require.NotNilf(t, err, "expected error on forbidden action")
+
+	ttErr, ok := err.(Error)
+	require.Equalf(t, ok, true, "error is built from a Tarantool error")
+
+	expected := BoxError{
+		Type:  "AccessDeniedError",
+		File:  "/__w/sdk/sdk/tarantool-2.10/tarantool/src/box/func.c",
+		Line:  uint64(535),
```
Comment (oleg-jukovec) [inline tarantool_test.go:3230]: The file and the line may change in the future.

File: box_error.go (hunk 19 above)
Comment (DifferentialOrange) [reply box_error.go:272]: Yeah, it seems I need to replace this one with `make`.

File: errors.go (hunk 23 above)
Comment (DifferentialOrange) [reply errors.go:9]: > err.Code == err.ExtendedInfo.Code?
> err.Msg == err.ExtendedInfo.Msg?

It is expected to (but documentation doesn't say anything).

> Does it make sense?
I think so. Should we start to return BoxError instead of Error since `BoxError` implements Error interface? I think type casts may start to fail.

File: tarantool_test.go (hunk 24 above)
Comment (DifferentialOrange) [reply tarantool_test.go:3230]: Yeah, I already have said that `CheckEqualBoxErrors` does not assert exact line and string (https://github.com/tarantool/go-tarantool/pull/229#discussion_r1033435847). But I think its better to have this backtrace + line here as an example of real BoxError.

In `box_error_test.go`, these fields are used on local encode/decode so everything should be alright. In any place where we work with an error from Tarantool instance, we do not assert the file and the line to be exactly equal to ones in example.

File: box_error.go (hunk 19 above)
Comment (DifferentialOrange) [reply box_error.go:272]: Wait a second, `make` is a tool to create slices.

Based on [this](https://stackoverflow.com/questions/31786937/are-all-the-variables-in-go-allocated-on-heap) stackoverflow answer, this approach seems ok:

> [There's a very clear answer to that question in the FAQ](https://golang.org/doc/faq#stack_or_heap):
> 
>     How do I know whether a variable is allocated on the heap or the stack?
> 
>     From a correctness standpoint, you don't need to know. Each variable in Go exists as long as there are references to it. The storage location chosen by the implementation is irrelevant to the semantics of the language.
> 
>     The storage location does have an effect on writing efficient programs. When possible, the Go compilers will allocate variables that are local to a function in that function's stack frame. However, if the compiler cannot prove that the variable is not referenced after the function returns, then the compiler must allocate the variable on the garbage-collected heap to avoid dangling pointer errors. Also, if a local variable is very large, it might make more sense to store it on the heap rather than the stack.
> 
>     In the current compilers, if a variable has its address taken, that variable is a candidate for allocation on the heap. However, a basic escape analysis recognizes some cases when such variables will not live past the return from the function and can reside on the stack.
> 
> TLDR: You shouldn't care. Go takes care of allocation for you.

File: box_error.go (hunk 20 above)
Comment (DifferentialOrange) [reply box_error.go]: Fixed

File: box_error_test.go (hunk 21 above)
Comment (DifferentialOrange) [reply box_error_test.go]: Fixed

File: box_error.go (hunk 18 above)
Comment (DifferentialOrange) [reply box_error.go]: Added error on empty stack decode and nil value encode.

File: const.go (hunk 22 above)
Comment (DifferentialOrange) [reply const.go]: Moved

File: box_error.go (hunk 19 above)
Comment (oleg-jukovec) [reply box_error.go:282]: https://go.dev/play/p/rrQdeadMzPJ :

```Go
import "fmt"

type foo struct {
	val int
}

func (a *foo) foo() {
	if a == nil {
		fmt.Println("a == nil")
		a = &foo{0}
		*a = foo{1}
	}
}

func main() {
	var bar *foo
	bar.foo()
	fmt.Println(bar)
}
// Output:
// a == nil
// <nil>
```

File: errors.go (hunk 23 above)
Comment (oleg-jukovec) [reply errors.go:9]: Oh, we can not do it anyway (field and method with the same name Error):

https://go.dev/play/p/-xbX5TYbgX5

So don't worry about it.

File: tarantool_test.go (hunk 24 above)
Comment (oleg-jukovec) [reply tarantool_test.go:3230]: Oh, sorry, I forgot it. Maybe then do not specify the values in the tests (or add a short note) so that it does not confuse in the future?

File: const.go [hunk 25]
```diff
@@ -35,13 +35,14 @@ const (
 	KeyExpression   = 0x27
 	KeyDefTuple     = 0x28
 	KeyData         = 0x30
-	KeyError        = 0x31
+	KeyError24      = 0x31 /* Error in pre-2.4 format */
```
Comment (oleg-jukovec) [inline const.go:38]:
```suggestion
	KeyError24      = 0x31 /* Error in pre-2.4 format. */
```

File: test_helpers/utils.go [hunk 26]
```diff
@@ -120,3 +121,75 @@ func SkipIfIdSupported(t *testing.T) {
...
+		t.Skip("Skipping test for Tarantool without error extended info support")
+	}
+}
+
+// SkipIfErrorExtendedInfoUnsupported skips test run if Tarantool without
+// MP_ERROR type over iproto support is used.
+func SkipIfErrorMessagePackTypeUnsupported(t *testing.T) {
+	t.Helper()
+
+	// Tarantool error type over MessagePack supported only since 2.10.0 version
```
Comment (oleg-jukovec) [inline test_helpers/utils.go:185]:
```suggestion
	// Tarantool error type over MessagePack supported only since 2.10.0 version.
```

**Review by oleg-jukovec — APPROVED** [review]
Thanks for the patch! I finally checked all dots at the ends of the comments. =)
It remains to resolve a couple of non-critical conversations.

File: box_error.go (hunk 19 above)
Comment (DifferentialOrange) [reply box_error.go:282]: I'm confused. So how does one fills the pointer value?

Comment (DifferentialOrange) [reply box_error.go:282]: Or is it simply impossible with nil pointer? The question is "What should I do if one had called UnmarshalMsgpack for a nil pointer?" and before I had added this line of code the answer was "crash with segmentation fault".

Comment (DifferentialOrange) [reply box_error.go]: Reworked to panic for `nil` pointer.

File: tarantool_test.go (hunk 24 above)
Comment (DifferentialOrange) [reply tarantool_test.go:3238]: Added

File: const.go (hunk 25 above)
Comment (DifferentialOrange) [reply const.go]: Fixed

File: test_helpers/utils.go (hunk 26 above)
Comment (DifferentialOrange) [reply test_helpers/utils.go]: Fixed

File: box_error.go (hunk 19 above)
Comment (DifferentialOrange) [reply box_error.go]: I think explicit panic is better, segfault is always dissatisfying.

**Review by AnaNek — APPROVED** [review]
LGTM.

File: CHANGELOG.md [hunk 27]
```diff
@@ -11,6 +11,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 ### Added
 
 - Support iproto feature discovery (#120).
+- Support errors extended information (#209).
+- Error type support in MessagePack (#209).
```
Comment (oleg-jukovec) [inline CHANGELOG.md:15]:
```suggestion
- Support errors extended information (#209)
- Error type support in MessagePack (#209)
```

**Review by oleg-jukovec — COMMENTED** [review]
You could also cherry-pick the commit:
https://github.com/tarantool/go-tarantool/pull/230/commits/8cb64890ede15748dafc21f19e2944686f90db58

**Comment by DifferentialOrange** [comment]
> You could also cherry-pick the commit: [8cb6489](https://github.com/tarantool/go-tarantool/commit/8cb64890ede15748dafc21f19e2944686f90db58)

Cherry-picked.

What's about https://github.com/tarantool/setup-tarantool/issues/37 ?

**Comment by DifferentialOrange** [comment]
> quickfix ubuntu-latest -> ubuntu-22.04

Quickfix is `ubuntu-latest` -> `ubuntu-20.04`

**Comment by oleg-jukovec** [comment]
> > quickfix ubuntu-latest -> ubuntu-22.04
> 
> Quickfix is `ubuntu-latest` -> `ubuntu-20.04`

Thank you, it works:

https://github.com/tarantool/go-tarantool/pull/230/commits/cc0398952d1afb1d691dedc5d27c432c3bf06c6c
